src/boot/efi/bcd.c

```diff
@@ -0,0 +1,286 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <efi.h>
+#include "macro-fundamental.h"
+#include "util.h"
+
+enum {
+        SIG_BASE_BLOCK = 1718052210, /* regf */
+        SIG_KEY = 27502,             /* nk */
+        SIG_SUBKEY_FAST = 26220,     /* lf */
+        SIG_KEY_VALUE = 27510,       /* vk */
+};
+
+enum {
+        REG_SZ = 1,
+        REG_MULTI_SZ = 7,
+};
+
+typedef struct {
+        UINT32 sig;
+        UINT32 primary_seqnum;
+        UINT32 secondary_seqnum;
+        UINT64 _pad1;
+        UINT32 version_major;
+        UINT32 version_minor;
+        UINT32 type;
+        UINT32 _pad2;
+        UINT32 root_cell_offset;
+        UINT64 _pad3[507];
+} _packed_ BaseBlock;
+assert_cc(sizeof(BaseBlock) == 4096);
+assert_cc(offsetof(BaseBlock, sig) == 0);
+assert_cc(offsetof(BaseBlock, primary_seqnum) == 4);
+assert_cc(offsetof(BaseBlock, secondary_seqnum) == 8);
+assert_cc(offsetof(BaseBlock, version_major) == 20);
+assert_cc(offsetof(BaseBlock, version_minor) == 24);
+assert_cc(offsetof(BaseBlock, type) == 28);
+assert_cc(offsetof(BaseBlock, root_cell_offset) == 36);
+
+/* All offsets are relative to the base block and technically point to a hive
+ * cell struct. But for our usecase we don't need to bother about that one,
+ * so skip over the cell_size UINT32. */
+#define HIVE_CELL_OFFSET (sizeof(BaseBlock) + 4)
+
+typedef struct {
+        UINT16 sig;
+        UINT16 _pad1[13];
+        UINT32 subkeys_offset;
+        UINT32 _pad2;
+        UINT32 n_key_values;
+        UINT32 key_values_offset;
+        UINT32 _pad3[7];
+        UINT16 key_name_len;
+        UINT16 _pad4;
+        CHAR8 key_name[];
+} _packed_ Key;
+assert_cc(offsetof(Key, sig) == 0);
+assert_cc(offsetof(Key, subkeys_offset) == 28);
+assert_cc(offsetof(Key, n_key_values) == 36);
+assert_cc(offsetof(Key, key_values_offset) == 40);
+assert_cc(offsetof(Key, key_name_len) == 72);
+assert_cc(offsetof(Key, key_name) == 76);
+
+typedef struct {
+        UINT16 sig;
+        UINT16 n_entries;
+        struct SubkeyFastEntry {
+                UINT32 key_offset;
+                CHAR8 name_hint[4];
+        } _packed_ entries[];
+} _packed_ SubkeyFast;
+assert_cc(offsetof(SubkeyFast, sig) == 0);
+assert_cc(offsetof(SubkeyFast, n_entries) == 2);
+assert_cc(offsetof(SubkeyFast, entries) == 4);
+
+typedef struct {
+        UINT16 sig;
+        UINT16 name_len;
+        UINT32 data_size;
+        UINT32 data_offset;
+        UINT32 data_type;
+        UINT32 _pad;
+        CHAR8 name[];
+} _packed_ KeyValue;
+assert_cc(offsetof(KeyValue, sig) == 0);
+assert_cc(offsetof(KeyValue, name_len) == 2);
+assert_cc(offsetof(KeyValue, data_size) == 4);
+assert_cc(offsetof(KeyValue, data_offset) == 8);
+assert_cc(offsetof(KeyValue, data_type) == 12);
+assert_cc(offsetof(KeyValue, name) == 20);
+
+static Key *get_key(const UINT8 *bcd, UINT32 bcd_len, UINT32 offset, const CHAR8 *name);
```

Comment:
I know this style is widely used, but please consider moving the star adjacent to the return base type:
```C
static Key* get_key(…);
```
I think that's more readable, because we're returning a `Key*` and now it's all in one place.

--

Since the input data is `const`, this should be `const Key*` too, here and below.